src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -0,0 +1,99 @@
+// <copyright file="EventCountersMetrics.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Diagnostics.Tracing;
+using System.Reflection;
+
+namespace OpenTelemetry.Instrumentation.EventCounters;
+
+/// <summary>
+/// .NET EventCounters Instrumentation.
+/// </summary>
+internal class EventCountersMetrics : EventListener
+{
+    internal static readonly AssemblyName AssemblyName = typeof(EventCountersMetrics).Assembly.GetName();
+    internal static readonly Meter MeterInstance = new(AssemblyName.Name, AssemblyName.Version.ToString());
+
+    private readonly EventCountersInstrumentationOptions options;
+    private readonly ConcurrentBag<EventSource> preInitEventSources = new();
+    private readonly ConcurrentDictionary<Tuple<string, string>, Instrument> instruments = new();
+    private readonly ConcurrentDictionary<Tuple<string, string>, double> values = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="EventCountersMetrics"/> class.
+    /// </summary>
+    /// <param name="options">The options to define the metrics.</param>
+    public EventCountersMetrics(EventCountersInstrumentationOptions options)
+    {
+        this.options = options;
```

Comment:
I think they could change it, but I also think you'd be well within your rights to say that is undefined behavior if they do:
```C#
var globalOptions;
builder.AddEventCounterInstrumentation(opts =>
{
    globalOptions = opts;
}).Build();
globalOptions.RefreshIntervalSecs = 1_000;
```
If you want to be very defensive you can copy the options in the constructor so that you never again refer to the mutable options object that was exposed in the callback.

Reply:
In general, OpenTelemetry .NET tries to be super conservative/defensive, `options` are only used during the ctor (and if someone trying to change the options while the ctor is running, it's considered their fault of introducing a race condition), all the necessary information in the `options` will be captured as a snapshot, so later modifications on the options won't affect anything that has been created already.

https://github.com/open-telemetry/opentelemetry-dotnet/blob/ed0450ef0ef5178cb29cb8908c3ff120a053caad/src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs#L47

Given https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/620#discussion_r978138219, I guess it's probably okay for now.